### DISCUSSION TITLE
Incremental source generators: how to deal with class hierarchies

### DISCUSSION THREAD

**Discussion by canton7** [dotnet/roslyn#60273] [Q&A]
I'm reading the [incremental source generator](https://github.com/dotnet/roslyn/blob/main/docs/features/incremental-generators.md) documentation, with a view to seeing whether it's viable to port one of my source generators.

My source generator needs to deal with class hierarchies: the user can define one partial type `Derived` which derives from another partial type `Base`, and I need to generate partial implementations for both. The catch is that they're not independent: the code generated for `Derived` depends on the code generated for `Base`.

In a non-incremental SG this is pretty straightforward: when I come to analyze a type, I first make sure that I have analyzed all of its base types, [like this](https://github.com/canton7/PropertyChanged.SourceGenerator/blob/cb9deb2f0ca15984f3e34f147c9c9e525ba1b5fa/src/PropertyChanged.SourceGenerator/Analysis/Analyser.cs#L66).

However, it's not clear how I'd do the same using an incremental SG. Ideally, I'd like the analysis of `Derived` to be re-run whenever either `Derived` changes, or the results of analysing `Base` change. However, there doesn't seem to be a way of declaring this: as I understand it, the directed graph of actions needs to be set in stone when the SG is initialised.

Since the analysis of `Base` is idempotent (two independent analyses will produce the same output), it might be possible to re-analyse `Base` whenever we do the analysis of `Derived` (so when `Base` changes we'll analyze `Base` as part of producing output for `Base`, and then analyze it again as part of producing output for `Derived`). However, this doesn't scale well to large class hierarchies (the number of analyses we need to perform progresses as the triangular numbers), and anyway there doesn't seem to be a way of declaring this "When `Base` changes, re-analyze `Derived`" dependency using the incremental SG API.

The fundamental problem is that the directed graph of dependencies provided by the incremental SG API is set in stone at initialisation. However, I need to be able to construct dependencies dynamically which reflect class hierarchies, and this doesn't seem to be possible.

What's the right way of approaching this problem using incremental SGs?

**Comment by jasonmalinowski** [comment]
So what sort of analysis are you doing on Base/Derived, and what does the your resulting generation look like?

**Reply by canton7** [reply]
This is an `INotifyPropertyChanged` generator -- [PropertyChanged.SourceGenerator](https://github.com/canton7/PropertyChanged.SourceGenerator).

One example is that the decision on whether to generate an `OnPropertyChanged` method (which raises the `PropertyChanged` event) in `Derived` depends on whether `Base` already defines it (or, indeed, whether the source generator has decided to generate it in `Base`). If `Base` doesn't implement INPC, then the source generator needs to implement it on `Derived`. Alternatively, if `Base` already implements `INPC`, or the source generator has determine that it needs to implement it on `Base`, then it shouldn't also implement it on `Derived`.

Another example is that the user can configure the source generator to raise a `PropertyChanged` event for a property on `Derived` when a property on `Base` changes, e.g. if the value of the property on `Derived` depends on the value of a property on `Base`. They can do this by putting e.g. `[DependsOn("BaseProperty")]` on the property on `Derived`. For this to work, the source generator has to know whether `BaseProperty` actually exists on `Base`, either because the user has defined it, or because the source generator has decided to generate it.

Another is that the source generate can be configured to set a particular property to `true` whenever any generated property is set, [as documented here](https://github.com/canton7/PropertyChanged.SourceGenerator#change-tracking-with-ischanged). This can be used by the user to track whether any modifications have been made to a model. When implementing properties on `Derived`, the source generator needs to know whether the base class has a property decorated with `[IsChanged]`.

In all of these cases, a change to `Base` will affect the code which needs to be generated for `Derived`.

Does that help?

**Reply by canton7** [reply]
> and what does the your resulting generation look like?

The analysis produced for a type [is here](https://github.com/canton7/PropertyChanged.SourceGenerator/blob/cb9deb2f0ca15984f3e34f147c9c9e525ba1b5fa/src/PropertyChanged.SourceGenerator/Analysis/TypeAnalysis.cs). The types it references are defined in files with the same name [in this folder](https://github.com/canton7/PropertyChanged.SourceGenerator/tree/cb9deb2f0ca15984f3e34f147c9c9e525ba1b5fa/src/PropertyChanged.SourceGenerator/Analysis).

**Reply by Sergio0694** [reply]
We're handling something similar in the MVVM Toolkit source generator. What we're doing is we also inspect the base type(s) and then determine what will the generator generator for that, and act as if that code had already been generated. Essentially, we pretend source generated code being generated concurrently was already present. As an example, if we need to check that the type has INPC support, we check whether it inherits from `ObservableObject` or whether any of the base types is annotated with `[ObservableObject]` or `[INotifyPropertyChanged]`. If that's the case, we rely on the fact that that generator will inject the corresponding code, and then just assume it's already there for the purposes of the current generation target. Does that help? 🙂

**Reply by canton7** [reply]
That seems to be the paragraph here?

> Since the analysis of Base is idempotent (two independent analyses will produce the same output), it might be possible to re-analyse Base whenever we do the analysis of Derived (so when Base changes we'll analyze Base as part of producing output for Base, and then analyze it again as part of producing output for Derived). However, this doesn't scale well to large class hierarchies (the number of analyses we need to perform progresses as the triangular numbers), and anyway there doesn't seem to be a way of declaring this "When Base changes, re-analyze Derived" dependency using the incremental SG API.

i.e. you re-analyse all of the base types for each derived type that you analyse?

Also, how do you declare that the `Derived` analysis needs to be re-run when the `Base` class changes?

**Reply by Sergio0694** [reply]
Ah, sorry, missed that one 😅

> "you re-analyse all of the base types for each derived type that you analyse?"

You don't really have to think about it thaat way. You re-analyze things whenever `ForAttributeWithMetadataName` calls you, which is roughly once per compilation. That doesn't change. The thing though is that from there you can build your incremental model with whatever info you need for generation, so all steps following that initial transform can be cached.

**Reply by canton7** [reply]
Thanks for your time.

Let's take an example:

 - `PropertyChangedBase` is defined in an external assembly, and contains the method `OnPropertyChanged`.
 - `Base` and `Derived` are defined in the user's assembly, which the SG runs on.

When the SG analyses `Base`, it also looks at `PropertyChangedBase`, and discovers the `OnPropertyChanged` method. When the SG analyses `Derived` it re-uses the analysis which it performed when analysing `Base`.

Is there a way to represent this with incremental SGs? The way that I understand them, I can identify `Base` and `Derived` individually, but I don't have a way of feeding just the results of the `Base` analysis into the code which analyses `Derived`. This means that I need to re-analyse `PropertyChangedBase` whenever `Derived` changes?

**Reply by Sergio0694** [reply]
Correct, each transform would run the analysis it needs to build its own incremental model. They don't communicate with each other.

**Reply by canton7** [reply]
OK that's a shame. Thanks a lot for you help!
